### ISSUE TITLE
Allow download of plugins and policy packs in advance of a `preview` or `up`

### ISSUE THREAD

**Issue by ghost** [pulumi/pulumi#9278]
## Allow download of plugins and policy packs in advance of a `preview` or `up`

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

We package and run our Pulumi programs in a very similar way to our applications. We bundle them into Docker containers during our CI process and then run the relevant commands (typically just `pulumi preview` and `pulumi up`) within those containers at deployment time to deploy our infrastructure.

In order to make this run efficiently I'd like to propose a few additional commands in the Pulumi CLI application to ensure the containers have everything they need to run. The build/test/linting steps are already pretty easy but currently we run into two problems with other dependencies:

- CrossGuard Policy Packs
- Provider plugins

When we run `pulumi up` from within these containers it spends a non-trivial amount of time downloading policy packs and plugins. The former actually caused a particularly frustrating problem for us a few months back - there was an npmjs.com outage that lasted an hour or so and during that time we were unable to use our deployment pipelines (installing policy packs required some other npm dependencies and we were unable to get Pulumi to use a registry proxy etc).

There are _some_ workarounds to this but they're pretty clunky:

Plugins:
- `pulumi plugin install [KIND NAME [VERSION]]` however this requires you to do some work to figure out which version you need and/or which versions are available. For example, write/run a script that queries https://api.github.com/repos/pulumi/pulumi/releases/latest for relevant versions and then loops through plugin names.

Policy Packs:
- Running a `pulumi preview` either before (or during) the build process and copying the downloaded policy packs into your image. This works but gets a little bit weird depending on how you approach it. Now you need cloud provider credentials at build time, or need to run a preview outside of a docker build context and then copy in the downloaded policy packs (and deal with any relevant binary dependencies properly) etc

## Proposed Solution

Ideally what I'd like to see is just two additional commands added to the Pulumi CLI that download the relevant plugins and policies for the current project. I'm not too fussed on exactly how it gets implemented but here's roughly what I had in mind:

`pulumi plugin install` with no additional arguments simply runs through the normal logic that occurs as part of an `up` or `preview` and downloads any missing provider plugins.

`pulumi policy install` that does the same as above but for policy packs.

If, at runtime the policies or available plugins have changed, they can/should be downloaded as normal, but at least this way in the vast majority of use cases the container will have everything it needs to be able to complete a `preview`/`run`.

Unsure how feasible these solutions are but I would be happy to participate in writing whatever functionality is agreed upon.

### Affected area/feature

- Pulumi CLI application

**Comment by justinvp** [comment]
> `pulumi plugin install` with no additional arguments

Note that `pulumi plugin install` w/ no additional arguments largely already does this today. It asks the language host for the current project for required plugins and installs them. What it doesn't currently do is also check for any plugins needed in the snapshot, but I don't see any reason why we couldn't make it do that as well.

**Comment by lukemundy-vgw** [comment]
yeah in most cases the `npm ci` part also installs the relevant plugins (removing the need to install plugins). I'll have to go back an identify where this didn't do what i expected, its possible the plugins part is no longer a problem.

The main inconvenience (for me at least) that motivated me to open this issue is the policy packs.

**Comment by Frassle** [comment]
Looked into this and it's a bit awkward because unlike `pulumi plugin install` which can ask the Project what plugins are needed, policy packs are assigned by Stacks so `pulumi policy install` would have to install all policies in the organisation (which it looks like we can get from "/orgs/{orgName}/policypacks")

**Comment by lukemundy-vgw** [comment]
would it work if we could pass some arguments to a hypothetical `pulumi policy install` command? either a list of stacks (and it would download the relevant policies for that stack) or a list of policy pack names?

In my use case the policies don't change often and generally the same policies are applied across all stacks in the org (with a few exceptions) so having to pass stacks or policy names to the command would suffice.

**Comment by aureq** [comment]
@justinvp @mikhailshilkov what's the current status for this request please?

**Comment by 88lexd** [comment]
This would be nice to speed up pulumi deployments in a pipeline using a known image without having to install dependencies every time

**Comment by aaomidi** [comment]
The annoying thing about this is that downloading the plugins at runtime can sometimes, outright, fail because of the rate-limiting in the package registrars. This is especially problematic if you've taken Pulumi as a runtime dependency...

**Comment by igoragoli** [comment]
For readers who want a quick solution for projects that use only a few plugins: adding a few `pulumi plugin install` lines to your containers' Dockerfiles is fine.
